hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java

```diff
@@ -364,27 +365,27 @@ public boolean hasTargetArtifactAssigned(final long targetId, final String sha1H
     }
 
     @Override
-    @Transactional
-    @Retryable(include = {
-            ConcurrencyFailureException.class }, maxAttempts = Constants.TX_RT_MAX, backoff = @Backoff(delay = Constants.TX_RT_DELAY))
-    public Target findOrRegisterTargetIfItDoesNotexist(final String controllerId, final URI address) {
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    @Retryable(include = ConcurrencyFailureException.class, exclude = EntityAlreadyExistsException.class,
```

Comment:
Should we exclude the spicific `EntityAlreadyExistsException` or the generic `DataIntegrityViolationException`?
Also, if the ConcurrencyFailureException is thrown after max retries, where is it handled? and why do we have to include it in the retry?

Reply:
Current solution is totally fine. `ConcurrencyFailureException` is in place to cover optimistic locking issues. If max retries are not sufficient you can't really fix it so it will be reported to the user